cmd/README.md

```diff
@@ -0,0 +1,639 @@
+- [Layout in Memory](https://docs.soliditylang.org/en/v0.8.10/internals/layout_in_memory.html)
+- [Layout of Call Data](https://docs.soliditylang.org/en/v0.8.10/internals/layout_in_calldata.html)
+- [Contract ABI Specification](https://docs.soliditylang.org/en/v0.8.10/abi-spec.html)
+- [Customizing the EVM with Stateful Precompiles](https://medium.com/avalancheavax/customizing-the-evm-with-stateful-precompiles-f44a34f39efd)
+- [Precompiles in Solidity](https://medium.com/@rbkhmrcr/precompiles-solidity-e5d29bd428c4)
+- [Customizing the EVM with Stateful Precompiles](https://medium.com/avalancheavax/customizing-the-evm-with-stateful-precompiles-f44a34f39efd)
+
+ 
+## Tutorial
+
+Let's start by creating the Solidity interface that we want to implement. We can put this  in `./contract-examples/contracts`
+
+```
+// (c) 2022-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.8.0;
+
+interface IHelloWorld {
+ // sayHello returns the string located at [key]
+  function sayHello() external returns (string calldata);
+
+// setGreeting sets the string located at [key]
+  function setGreeting(string calldata response) external;
+}
+```
+
+Now we have an interface that our precompile can implement!
+Let's create an [abi](https://docs.soliditylang.org/en/v0.8.13/abi-spec.html#:~:text=Contract%20ABI%20Specification-,Basic%20Design,as%20described%20in%20this%20specification.) of our solidity code.
+
+In the same `./contract-examples/contracts` directory, let's download solc and run
+```
+brew update
+brew tap ethereum/ethereum
+brew install solidity
+
+solc --abi IHelloWorld.sol -o .
+```
+
+This spits out the abi code. Let's move it into a brand new folder in 
+`./contract-examples/contracts/contract-abis`. 
+
+IHelloWorld.abi
+
+```
+[{"inputs":[],"name":"sayHello","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"string","name":"recipient","type":"string"}],"name":"setGreeting","outputs":[],"stateMutability":"nonpayable","type":"function"}]
+```
+
+
+## Precompile tool
+
+The precompile tool can take in 4 arguments. 
+
+
+### `--abi` (string):
+
+It needs an ABI input so it can bind it to the precompile template.
+
+`--abi ./contract-examples/contracts/contract-abis/IHelloWorld.abi`
+
+### `--type `(string):
+
+It takes in a type which it uses as a struct name for the precompile. This is optional 
+and will default to the abi name. 
+
+`--type HelloWorld`
+
+### `--pkg` (string):
+
+It also takes in a pkg, which is the package name to generate the precompile into. 
+This is optional and it defaults to "precompile".
+
+`--pkg precompile` 
+
+### `--out `(string):
+
+Finally it can take in an out which is the path and name of the output file of the generated precompile. 
+
```

Comment:
We should separate this documentation from the tutorial itself, so that there is the documentation of what the tool does and a separate tutorial that walks the user through how to use it.

Reply:
I find this helpful here, because it allows for the user on how to customize their command. Can we keep it here for now?

Reply:
I agree, if we aim to put this into docs repo we should split this and put PrecompileGen tool under Tools section in https://docs.avax.network/subnets. Then we can put a link here instead.

Alternatively for now we can (and probably should) move this section under `cmd/precompilegen/README.md`, and link here.

Reply:
Let's put this into the docs repo and link to it or give a full command with the fill in the blanks of the command arguments here, a tutorial is not the best place to provide reference documentation like this. I actually found @dasconnor 's recent tweet to be great on this point: https://documentation.divio.com/introduction/